crates/ruff/src/rules/perflint/rules/unnecessary_list_cast.rs

```diff
@@ -0,0 +1,100 @@
+use crate::registry::AsRule;
+
+/// ## What it does
+/// Checks for explicit usages of `list()` on an iterable to iterate over them in a for-loop
+///
+/// ## Why is this bad?
+/// Using a `list()` call to eagerly iterate over an already iterable type is inefficient as a
+/// second list iterator is created, after first iterating the value:
+///
+/// ## Example
+/// ```python
+/// items = (1, 2, 3)
+/// for i in list(items):
+///     print(i)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// items = (1, 2, 3)
+/// for i in items:
+///     print(i)
+/// ```
+#[violation]
+pub struct UnnecessaryListCast;
+
+impl AlwaysAutofixableViolation for UnnecessaryListCast {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Do not cast an iterable to `list` before iterating over it")
+    }
+
+    fn autofix_title(&self) -> String {
+        format!("Remove `list()` cast")
+    }
+}
+
+/// PERF101
+pub(crate) fn unnecessary_list_cast(checker: &mut Checker, iter: &Expr) {
+    let Expr::Call(ast::ExprCall{ func, args, range, ..}) = iter else {
+        return;
+    };
+
+    if args.is_empty() {
+        return;
+    }
+
+    if let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() {
+        if id != "list" || !checker.semantic().is_builtin("list") {
+            return;
+        }
+    };
+
+    match &args[0] {
+        Expr::Tuple(_) | Expr::List(_) | Expr::Set(_) => fix_incorrect_list_cast(checker, *range),
+        Expr::Name(ast::ExprName { id, .. }) => {
+            let scope = checker.semantic().scope();
+            if let Some(binding_id) = scope.get(id) {
+                let binding = &checker.semantic().bindings[binding_id];
+                if binding.kind.is_assignment() || binding.kind.is_named_expr_assignment() {
+                    if let Some(parent_id) = binding.source {
+                        let parent = checker.semantic().stmts[parent_id];
+                        match parent {
+                            Stmt::Assign(ast::StmtAssign { value, .. })
+                            | Stmt::AnnAssign(ast::StmtAnnAssign {
+                                value: Some(value), ..
+                            }) => match value.as_ref() {
+                                Expr::Tuple(_) | Expr::List(_) | Expr::Set(_) => {
+                                    fix_incorrect_list_cast(checker, *range);
+                                }
+                                _ => {}
+                            },
+                            _ => (),
+                        }
+                    }
+                }
+            }
+        }
+        _ => {}
+    }
+}
+
+fn fix_incorrect_list_cast(checker: &mut Checker, range: TextRange) {
+    let mut diagnostic = Diagnostic::new(UnnecessaryListCast, range);
+    if checker.patch(diagnostic.kind.rule()) {
+        diagnostic.set_fix(Fix::automatic_edits(
+            Edit::deletion(range.start(), range.start() + TextSize::from(5)),
```

Comment:
If there's a more elegant way to extract the string representation of the value in the `list()` call to replace the whole thing instead of messing about with set ranges, I'm happy for the input.

Reply:
Can you use the start of the tuple? It has the downside that it truncates comments but produces a correct fix if there's a space (or even comment between list and the (

Reply:
Sure! I'm not sure I exactly understand what you mean, though. Can you maybe give an example that would break in this implementation that I can add to the fixture to check after I have made the tweak if I understood?

Reply:
Because the range being passed in now is already the range of the `iter: &Expr` iterator. And if you mean the iterable then I don't really see how that helps with comments (aside from the issue of getting the range from the iterable depending on whether it's a binding or literal etc.)

Reply:
Ah wait I think I get it. Do you mean taking the deletion range to be from the start of the iterator to the start of the iterable? 

Now that I'm looking it over that still leaves another scenario that would be unfixable namely:
```python
for bla in list(
    {1, 2, 3}
)
```

The most complete way would be to somehow extract whatever is in the `list()` call in its entirety, trim the whitespace, and replace the whole range of the iterator with that value. That would catch multiline args but would still leave the issue of comments in something like:

```python
for i in list(
    # bla bla
    {1, 2, 3}
): 
    pass
```

I'll see if I can figure something out tomorrow and otherwise I might leave the autofix for later

Reply:
You would probably need to use libCST if you want to retain comments. Many of Ruff's fixes do not handle comments.

I think it's otherwise fine to delete the ranges from the `list(` call expression start to the start of the iterable and from the end of the iterable to the end of the `list call`

```python
for bla in list(   { 1, 2, 3 }  ):
           ^^^^^^^^           ^^^
	...
```

Reply:
Ah in that case I'll just delete the comments. Thanks for the feedback, pushed change!